dan_layer/engine/src/runtime/tracker.rs

```diff
@@ -502,16 +502,38 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> StateTracke
     ) -> Result<FinalizeTracker, RuntimeError> {
         // Resolve the transfers to the fee pool resource and vault refunds
         let finalized_fee = self.finalize_fees(&mut substates_to_persist)?;
+
+        let transaction_receipt = substates_to_persist
+            .entry(SubstateAddress::TransactionReceipt(self.transaction_hash().into()))
+            .and_modify(|x| {
+                x.into_transaction_receipt().and_then(|x| {
+                    Some(TransactionReceipt {
+                        transaction_hash: x.transaction_hash,
+                        events: x.events.clone(),
+                        logs: x.logs.clone(),
+                        fee_receipt: Some(finalized_fee.clone()),
+                    })
+                });
+            })
+            .or_insert_with(|| panic!("Failed to get transaction receipt"));
         // events and logs
-        let events = self.take_events();
-        let logs = self.take_logs();
+        let events = transaction_receipt
+            .into_transaction_receipt()
+            .expect("Failed to get transaction receipt")
+            .events
+            .clone();
+        let logs = transaction_receipt
+            .into_transaction_receipt()
+            .expect("Failed to get transaction receipt")
+            .logs
+            .clone();
 
         // Finalise will always reset the state
         let state = self.take_working_state();
 
         let result = state
             .validate_finalized()
-            .and_then(|_| self.generate_substate_diff(state, substates_to_persist));
+            .and_then(|_| self.generate_substate_diff(state, substates_to_persist.clone()));
```

Comment:
This clone should be needed (?)
```suggestion
            .and_then(|_| self.generate_substate_diff(state, substates_to_persist));
```